### DISCUSSION TITLE
Separating GUI from Business logic

### DISCUSSION THREAD

**Discussion by bofdahof** [FreeCAD/FreeCAD#18241] [Discussions that will turn in to issues but need refining]
Throughout FreeCAD the GUI and Business logics are inextricably entwined in spaghetti.

IMHO it is the job of a gui component to display something, interact with the user, and report back to the business logic. The gui code should be TOTALLY unaware of the business logic and associated code.

Am I missing something? 

If not, how to begin to fix?

Unfortunately this board does not seem to attract much traffic, but will give it a shot ...

**Comment by luzpaz** [comment]
technical issues like this get more traffic on the original forum.

**Reply by bofdahof** [reply]
But they seem to get deleted there by moderator ..

**Reply by luzpaz** [reply]
Are you posting in the developer forum ?

**Reply by bofdahof** [reply]
@luzpaz I no longer see this topic in discussions?

**Comment by tritao** [comment]
@bofdahof Agree with you on this one, this is also something I'd like to see cleaned up in FreeCAD.

As to how to start to clean it up, it is indeed a difficult job that will require some previous discussion.

I think all GUI logic in C++ (besides the core widgets, I suppose) could be moved to Python, where it's much easier to modify and work with.

And some event bus or even the existing signal/slot system could be used for messaging between the C++ operations and Python GUI logic.

**Reply by kadet1090** [reply]
C++ gives use much greater abilities in terms of interaction with whole codebase. We want to move into QML soon and this should help with separation of concerns.

**Comment by bofdahof** [comment]
@tritao I did an experimental version in C++ of the 'First Use' screen. It certainly wasn't difficult or complex, just separation of concerns, but definitely a different way of thinking to what has become the norm in FreeCAD. Maybe will put up as draft PR just as demonstration.
